### PR TITLE
initial implementation of machinset controller

### PR DESCRIPTION
**Pull request by p0lyn0mial**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note

```

@kubernetes/kube-deploy-reviewers

### CODE REVIEW

**Comment by p0lyn0mial** [comment]
PTAL, as this is something we are going to push upstream, I'd like to have your opinions.

File: cluster-api/examples/manifests/machinesets.yaml [hunk 1]
```diff
@@ -0,0 +1,31 @@
+apiVersion: "cluster.k8s.io/v1alpha1"
+kind: MachineSet
+metadata:
+  name: my-first-machineset
+spec:
+  replicas: 3
+  template:
+    spec:
+      providerConfig: |
+        sshPublicKeys:
```
Comment (mrIncompetent) [inline cluster-api/examples/manifests/machinesets.yaml]: Please use a config form cluster-api-gcp/machines.yaml.template

File: cluster-api/pkg/controller/machineset/controller.go [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	glog.Info("starting machine controller")
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	glog.Info("shutting down machine controller")
```
Comment (mrIncompetent) [inline cluster-api/pkg/controller/machineset/controller.go]: machineset controller

File: cluster-api/pkg/controller/machineset/controller.go [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+// that are owned by the given machineSet (key).
+func (c *Controller) syncHandler(key string) error {
+	machineSet, err := c.machineSetsLister.Get(key)
+	if err != nil {
+		return err
+	}
+
+	var filteredMachines []*v1alpha1.Machine
+	{
+		allMachines, err := c.machineLister.List(labels.Everything())
```
Comment (mrIncompetent) [inline cluster-api/pkg/controller/machineset/controller.go]: the complete block should be moved into a function

File: cluster-api/pkg/controller/machineset/controller.go [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+			return err
+		}
+		if err = waitForMachineInCache(machine.Name, c.machineLister); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	if desiredReplicas < currentMachineCount {
+		glog.Infof("deleting a machine ( spec.replicas(%d) < currentMachineCount(%d) )", desiredReplicas, currentMachineCount)
```
Comment (mrIncompetent) [inline cluster-api/pkg/controller/machineset/controller.go]: Use `glog.V(2).Infof` instead

File: cluster-api/pkg/controller/machineset/controller.go [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+// syncReplicas essentially scales machine resources up and down.
+func (c *Controller) syncReplicas(machineSet *v1alpha1.MachineSet, machines []*v1alpha1.Machine) error {
+	currentMachineCount := int32(len(machines))
+	desiredReplicas := int32(0)
+	if machineSet.Spec.Replicas != nil {
+	 desiredReplicas = *machineSet.Spec.Replicas
+	}
+
+	if desiredReplicas > currentMachineCount {
+		glog.Infof("creating a machine ( spec.replicas(%d) > currentMachineCount(%d) )", desiredReplicas, currentMachineCount)
```
Comment (mrIncompetent) [inline cluster-api/pkg/controller/machineset/controller.go]: Use `glog.V(2).Infof` instead

**Review by mrIncompetent — CHANGES_REQUESTED** [review]
nice!

**Comment by p0lyn0mial** [comment]
@mrIncompetent thanks for your comments.

**Pull request closed without merging**